zebra-chain/src/orchard/keys.rs

```diff
@@ -0,0 +1,1008 @@
+                Ok(IncomingViewingKey {
+                    network: match hrp.as_str() {
+                        ivk_hrp::MAINNET => Network::Mainnet,
+                        _ => Network::Testnet,
+                    },
+                    scalar: pallas::Scalar::from_bytes(&scalar_bytes).unwrap(),
+                })
+            }
+            _ => Err(SerializationError::Parse("bech32 decoding error")),
+        }
+    }
+}
+
+impl PartialEq<[u8; 32]> for IncomingViewingKey {
+    fn eq(&self, other: &[u8; 32]) -> bool {
+        self.scalar.to_bytes() == *other
+    }
+}
+
+/// Magic human-readable strings used to identify what networks Orchard full
+/// viewing keys are associated with when encoded/decoded with bech32.
+///
+/// https://zips.z.cash/protocol/nu5.pdf#orchardfullviewingkeyencoding
+mod fvk_hrp {
+    pub const MAINNET: &str = "zviewo";
+    pub const TESTNET: &str = "zviewtestorchard";
+}
+
+/// Full Viewing Keys
+///
+/// Allows recognizing both incoming and outgoing notes without having
+/// spend authority.
+///
+/// For incoming viewing keys on the production network, the
+/// Human-Readable Part is “zviewo”. For incoming viewing keys on the
+/// test network, the Human-Readable Part is “zviewtestorchard”.
+///
+/// https://zips.z.cash/protocol/protocol.pdf#orchardfullviewingkeyencoding
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct FullViewingKey {
+    network: Network,
+    spend_validating_key: SpendValidatingKey,
+    nullifier_deriving_key: NullifierDerivingKey,
+    ivk_commit_randomness: IvkCommitRandomness,
+}
+
+// TODO: impl a From that accepts a Network?
+
+impl fmt::Debug for FullViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("FullViewingKey")
+            .field("network", &self.network)
+            .field("spend_validating_key", &self.spend_validating_key)
+            .field("nullifier_deriving_key", &self.nullifier_deriving_key)
+            .field("ivk_commit_randomness", &self.ivk_commit_randomness)
+            .finish()
+    }
+}
+
+impl fmt::Display for FullViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut bytes = io::Cursor::new(Vec::new());
+
+        let _ = bytes.write_all(&<[u8; 32]>::from(self.spend_validating_key));
+        let _ = bytes.write_all(&<[u8; 32]>::from(self.nullifier_deriving_key));
+        let _ = bytes.write_all(&<[u8; 32]>::from(self.ivk_commit_randomness));
+
+        let hrp = match self.network {
+            Network::Mainnet => fvk_hrp::MAINNET,
+            _ => fvk_hrp::TESTNET,
+        };
+
+        bech32::encode_to_fmt(f, hrp, bytes.get_ref().to_base32(), Variant::Bech32).unwrap()
+    }
+}
+
+impl From<SpendingKey> for FullViewingKey {
+    fn from(sk: SpendingKey) -> FullViewingKey {
+        let spend_authorizing_key = SpendAuthorizingKey::from(sk);
+
+        Self {
+            // TODO: handle setting the Network better.
+            network: Network::default(),
```

Comment:
Can we change this into a `FullViewingKey::from_spending` function that takes a `Network`?

(Maybe we should remove the default on `Network`, to avoid accidental mainnet-only implementations, they have happened in a bunch of Zebra code.)

Reply:
Done: 973246c4